src/parsers.rs

```diff
@@ -164,6 +164,20 @@ impl<'a> Parser<'a> {
             }
         };
 
+        // Check for an IPv4 mapped address
+        if *head_idx == 6 || use_tail && *tail_idx == 1 {
+            match self.try(|p| p.accept_ipv4_octets()) {
+                Some(octets) => {
+                    tail[*tail_idx] = ((octets[0] as u16) << 8) | (octets[1] as u16);
+                    *tail_idx += 1;
+                    tail[*tail_idx] = ((octets[2] as u16) << 8) | (octets[3] as u16);
+                    *tail_idx += 1;
+                    return Ok(());
+                },
+                None => ()
+            }
+        }
+
         match self.try(|p| p.accept_number(4, 0x10000, true)) {
```

Comment:
IMO at a first glance, I'd say the parsing of the ipv4 address should be a match arm here if the number is is `0xffff` and `*head_idx == 6 || use_tail && *tail_idx == 1`. This would also increase readability in my opinion.

Reply:
i thought about doing this, but it will cause a lot of change in the parser. before trying to parse a ipv4 address, i would have to `self.accept_char(b':')?;`. should the parser fail, the check down below looks ahead of `b':'` and errors if one is not present. i would also have to remove the `self.accept_char(b':')?;` in the `None` case for the double colon check.

all that being said, i see your point about readability. i will see what i can make better.

Reply:
ignore my last comment. wrapping it in `self.try()` works fine.

i did not make it a separate match arm. i wrote that at first but that caused me to duplicate the logic of the next 3 match arms because i need to do something with `part` still. instead, i incorporated it into the match arms for head and tail. i think it reads better now. let me know what you think.

Reply:
Nice work. The update is much more readable